zng/docs/zng-over-json.md

```diff
@@ -1,387 +1,396 @@
+so appropriate for structured data format like ZNG.
+That said, JSON can be used an encoding format for ZNG by mapping ZNG data
+onto a JSON-based protocol.  This allows clients like web apps or
+electron apps to receive and understand ZNG and, with the help of client
+libraries like [zealot](https://github.com/brimsec/brim/tree/master/zealot),
+to be enabled with rich, structured ZNG types that are implemented on top of
+the basic javascript types.
 
-This zson...
+In other words,
+because JSON objects do not have a deterministic column order nor does JSON
+in general have typing beyond the basics (i.e., strings, floating-point numbers,
+objects, arrays, and booleans), we decided to encode the ZNG data format with
+its embedded type model all in a layer above regular JSON.
 
+## ZJSON
+
+The format for representing ZNG in JSON is called ZJSON.
+Converting ZNG to ZJSON and back results in a complete and accurate
+restoration of the original ZNG.
+
+The ZJSON data model follows that of the underlying ZNG model by embedding
+type information in the stream: type definitions declare arbitrarily complex
+and nested data types, and values are sent referencing the type information
+recursively with small-integer type identifiers.
+
+Since ZNG steams are self describing and type information is embedded
+in the stream itself, the embedded types are likewise encoded in the
+ZJSON format.
+
+A ZJSON stream defined as a sequence of JSON objects where each object
+represents a ZNG value.  Each object includes an identifier that denotes
+its type, or "schema".  A schema generically refers to the type of the
+ZNG record that defined by a given JSON object.
+
+Each object contains the following fields:
+* `id` a small integer encoded as a javascript number indicating the schema that
+applies to this value,
+* `values` a JSON array of strings and arrays encoded as defined below,
+* `schema` an optional field encoding the type of this object's values
+and the schema of all subsequent values that have the same identifier, as
+defined below, and
+* `aliases` a JSON array of objects that define bindings between string names
+and ZNG types as defined below.
+
+The ID provides a mapping to a type so that future values in the stream may
+reference a schema by ID.  An implementation maintains a table to map schemas
+to types as it decodes values.  The ID's are scoped to the particular ZJSON
+data stream in which they are embedded and otherwise have no global persistence
+or meaning.
+
+The form of each value has the following JSON structure:
 ```
-#1:record[id:record[orig_h:addr,orig_p:port,resp_h:addr,resp_p:port],etc:string]
-1:[[192.168.1.1;80;192.168.2.22;6060;]foo;]
-1:[[192.168.1.2;80;192.168.2.21;5050;]bar;]
-1:[[192.168.1.3;80;192.168.2.20;4040;]hello;]
+{
+        id: <id>,
+        schema: <type>,
+        aliases: [ <alias1>, <alias2>, ... <aliasn> ],
+        values: [ <val> ... [ <val>, ... ] ... ]
```

Comment:
Nit: Put these in the same order in which they're defined above.

Reply:
The values depend on the aliases but ok.

Reply:
Would it make more sense to maintain this order here but change the order in which the fields are defined?